Typechecker crash when accessing struct fields with array type

## Description

Accessing a struct field that has an array type causes a nil pointer dereference in the typechecker's `inferMemberType` function.

## Reproduction

```ez
const TestResult struct {
    passed int
    bugs [string]
}

do main() {
    temp r TestResult = TestResult{passed: 1, bugs: {"bug1"}}

    // This crashes:
    temp bug_list [string] = r.bugs
}
```

## Error

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x...]

goroutine 1 [running]:
github.com/marshallburns/ez/pkg/typechecker.(*TypeChecker).inferMemberType(...)
    pkg/typechecker/typechecker.go:2215 +0x90
```

## Expected Behavior

Accessing struct fields with array types should work correctly.

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -438,7 +438,18 @@ func (tc *TypeChecker) checkStructDeclaration(node *ast.StructDeclaration) {
 		}
 
 		// Add field to struct type
-		fieldType, _ := tc.GetType(field.TypeName)
+		fieldType, ok := tc.GetType(field.TypeName)
+		if !ok {
+			// For array/map types, create a Type on-the-fly since they're not in the registry
+			if tc.isArrayType(field.TypeName) {
+				fieldType = &Type{Name: field.TypeName, Kind: ArrayType}
+			} else if tc.isMapType(field.TypeName) {
+				fieldType = &Type{Name: field.TypeName, Kind: MapType}
+			} else {
+				// This shouldn't happen since TypeExists passed, but be safe
+				continue
+			}
+		}
 		structType.Fields[field.Name.Value] = fieldType
 	}
 }

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -482,6 +482,57 @@ do main() {
 	assertNoErrors(t, tc)
 }
 
+func TestStructWithArrayField(t *testing.T) {
+	// Regression test for issue #336: typechecker crashed when accessing
+	// struct fields with array types due to nil pointer dereference
+	input := `
+const TestResult struct {
+	passed int
+	bugs [string]
+}
+
+do main() {
+	temp r TestResult = TestResult{passed: 1, bugs: {"bug1"}}
+	temp bug_list [string] = r.bugs
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestStructWithMapField(t *testing.T) {
+	// Ensure struct fields with map types also work correctly
+	input := `
+const Config struct {
+	name string
+	settings map[string:int]
+}
+
+do main() {
+	temp c Config = Config{name: "test", settings: {"a": 1}}
+	temp s map[string:int] = c.settings
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestStructWithNestedArrayField(t *testing.T) {
+	// Test nested array types in struct fields
+	input := `
+const Matrix struct {
+	rows [[int]]
+}
+
+do main() {
+	temp m Matrix = Matrix{rows: {{1, 2}, {3, 4}}}
+	temp r [[int]] = m.rows
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
 // ============================================================================
 // Enum Tests
 // ============================================================================

